registry/warming.go

```diff
@@ -0,0 +1,234 @@
+// Runs a daemon to continuously warm the registry cache.
+package registry
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+
+	"context"
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/weaveworks/flux"
+	"github.com/weaveworks/flux/registry/cache"
+	"strings"
+)
+
+const refreshWhenExpiryWithin = time.Minute
+
+type Warmer struct {
+	Logger        log.Logger
+	ClientFactory ClientFactory
+	Creds         Credentials
+	Expiry        time.Duration
+	Writer        cache.Writer
+	Reader        cache.Reader
+	Burst         int
+}
+
+// Continuously wait for a new repository to warm
+func (w *Warmer) Loop(stop <-chan struct{}, wg *sync.WaitGroup, warm <-chan flux.ImageID) {
+	defer wg.Done()
+
+	if w.Logger == nil || w.ClientFactory == nil || w.Expiry == 0 || w.Writer == nil || w.Reader == nil {
+		panic("registry.Warmer fields are nil")
+	}
+
+	for {
+		select {
+		case <-stop:
+			w.Logger.Log("stopping", "true")
+			return
+		case r := <-warm:
+			w.warm(r)
+		}
+	}
+}
+
+func (w *Warmer) warm(id flux.ImageID) {
```

Comment:
A general observation about this function: it is bloated because the cache API is too low level - IMO we should push the JSON marshalling and key extraction responsibilities down into the cache, since every user wants those things. Then you'd just have single operations here (e.g. 'put this in the cache') rather than making the key/marshalling the json/setting the thing